cosmos_rl/rollout/vllm_rollout/vllm_rollout_worker_async.py

```diff
@@ -0,0 +1,1469 @@
+                    )
+        except asyncio.TimeoutError:
+            # Timeout reached, return normally
+            pass
+
+    @torch.no_grad()
+    async def do_validation(self):
+        # submit payloads to scheduler
+        prompt_idxs: List[int] = []
+        validation_payloads: List[RLPayload] = []
+        # Do validation here
+        is_end = False
+        while True:
+            if not is_end:
+                is_end = self.request_new_prompts(
+                    self.val_batch_size,
+                    validation_step=self.current_step,
+                )
+
+            # wait all tasks are completed
+            while self.scheduler.completed_results() != self.val_batch_size:
+                await asyncio.sleep(0.1)
+
+            rollout_results = self.scheduler.get_all()
+            prompt_idxs.extend([p.idx for p in rollout_results])
+            validation_payloads.extend([p.payload for p in rollout_results])
+
+            if is_end and self.scheduler.is_idle():
+                break
+
+        # Clear the flag to indicate validation is done.
+        self.validation_flag.clear()
+        should_report = self.parallel_dims.tp_coord[0] == 0 and (
+            self.parallel_dims.pp_coord[0] == self.parallel_dims.pp_coord[1] - 1
+        )
+
+        if should_report:
+            self.reward_dispatcher.enqueue_rewards_cal(
+                validation_payloads, True, self.current_step, prompt_idxs
+            )
+            payloads, is_validation, current_step, empty = self.report_rollouts(
+                block=True
+            )
+            assert (
+                (is_validation and payloads is not None or payloads is None)
+                and (not empty or len(validation_payloads) == 0)
+            ), f"Payloads must be for validation if not empty {is_validation}, {payloads}, {empty}"
+            while not empty:
+                assert (
+                    is_validation or payloads is None
+                ), f"Payloads must be for validation if not empty {is_validation}, {payloads}, {empty}"
+                if payloads is not None:
+                    response = ValidationReportRequest(
+                        src_replica_name=self.replica_name,
+                        validation_step=current_step,
+                        prompt_idxs=[],
+                        payloads=payloads,
+                        is_end=True,
+                    )
+                    self.api_client.post_validation_report(response)
+                payloads, is_validation, current_step, empty = (
+                    self.reward_dispatcher.dequeue_rewards_cal()
+                )
+
+    @torch.no_grad()
+    async def do_generate(self):
+        logger.debug(f"[Rollout] generate start for rank {self.global_rank}")
+
+        # we enqueue tasks by batch in the main loop, so we need to collect all task results here.
+        await self.scheduler.wait_all_tasks_completed()
```

Comment:
This wait all tasks completed will drain the pipeline, which might lower the throughput? Could we don't do such wait and just collect the already ready rollouts only?